pkg/controller/nodeipam/ipam/range_allocator.go

```diff
@@ -228,99 +241,141 @@ func (r *rangeAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
 		klog.V(2).Infof("Node %v is already in a process of CIDR assignment.", node.Name)
 		return nil
 	}
-	if node.Spec.PodCIDR != "" {
-		return r.occupyCIDR(node)
+
+	if 0 < len(node.Spec.PodCIDRs) {
+		return r.occupyCIDRs(node)
 	}
-	podCIDR, err := r.cidrs.AllocateNext()
-	if err != nil {
-		r.removeNodeFromProcessing(node.Name)
-		nodeutil.RecordNodeStatusChange(r.recorder, node, "CIDRNotAvailable")
-		return fmt.Errorf("failed to allocate cidr: %v", err)
+	// allocate and queue the assignment
+	allocated := nodeReservedCIDRs{
+		nodeName:       node.Name,
+		allocatedCIDRs: make([]*net.IPNet, len(r.cidrSets)),
 	}
 
-	klog.V(4).Infof("Putting node %s with CIDR %s into the work queue", node.Name, podCIDR)
-	r.nodeCIDRUpdateChannel <- nodeAndCIDR{
-		nodeName: node.Name,
-		cidr:     podCIDR,
+	for idx := range r.cidrSets {
+		podCIDR, err := r.cidrSets[idx].AllocateNext()
+		if err != nil {
+			r.removeNodeFromProcessing(node.Name)
+			nodeutil.RecordNodeStatusChange(r.recorder, node, "CIDRNotAvailable")
+			return fmt.Errorf("failed to allocate cidr from cluster cidr at idx:%v: %v", idx, err)
+		}
+		allocated.allocatedCIDRs[idx] = podCIDR
 	}
+
+	//queue the assignement
+	klog.V(4).Infof("Putting node %s with CIDR %v into the work queue", node.Name, allocated.allocatedCIDRs)
+	r.nodeCIDRUpdateChannel <- allocated
 	return nil
 }
 
+// ReleaseCIDR marks node.podCIDRs[...] as unused in our tracked cidrSets
 func (r *rangeAllocator) ReleaseCIDR(node *v1.Node) error {
-	if node == nil || node.Spec.PodCIDR == "" {
+	if node == nil || 0 == len(node.Spec.PodCIDRs) {
 		return nil
 	}
-	_, podCIDR, err := net.ParseCIDR(node.Spec.PodCIDR)
-	if err != nil {
-		return fmt.Errorf("Failed to parse CIDR %s on Node %v: %v", node.Spec.PodCIDR, node.Name, err)
-	}
 
-	klog.V(4).Infof("release CIDR %s", node.Spec.PodCIDR)
-	if err = r.cidrs.Release(podCIDR); err != nil {
-		return fmt.Errorf("Error when releasing CIDR %v: %v", node.Spec.PodCIDR, err)
+	for idx, cidr := range node.Spec.PodCIDRs {
```

Comment:
Clarifying question: the order of the pod CIDRs is guaranteed by API? Couldn't find it in this PR probably just missed it

Reply:
I am not sure i fully understand the question. The order CIDRs is guaranteed, yes. It is the semantics apimachinery. This PR didn't do anything to enable it.

Reply:
So we assume the order of the address family stored in cidr sets is the same as `node.Spec.PodCIDRs`, [v4, v6]?

Reply:
We copy the order of cidrs as the user set them. if the user set `v4,v6` then node cidrs will be set (and guaranteed to always read as) `v4,v6` .. same for vice versa..